## Question 1: What is the relationship between various socio-economic factors and the share of female STEM graduates over time, and can these factors predict future trends in the proportion of women entering STEM fields?

In [1]:
#import packages

# general
import numpy as np
import pandas as pd
import time

# sklearn
# sklearn
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.model_selection import train_test_split, LeaveOneOut, KFold, cross_val_score
from sklearn.metrics import mean_squared_error
from sklearn.feature_selection import SequentialFeatureSelector
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import ElasticNet
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix, ConfusionMatrixDisplay
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso
from sklearn.metrics import classification_report
from sklearn.model_selection import GridSearchCV
from sklearn import svm
import pandas as pd
from sklearn.metrics import accuracy_score

# visualization
import seaborn as sns
from matplotlib import pyplot as plt
%matplotlib inline
import numpy as np
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt
import statsmodels.formula.api as smf

In [2]:
#load in the dataset
df = pd.read_csv('transformed_data.csv')
df.head()

Unnamed: 0  Year  \
0           0  -1.0   
1           1  -0.7   
2           2   0.1   
3           3   0.5   
4           4   0.6   

   A woman can be head of household in the same way as a man (1=yes; 0=no)  \
0                                                1.0                         
1                                                1.0                         
2                                                1.0                         
3                                                1.0                         
4                                                1.0                         

   A woman can choose where to live in the same way as a man (1=yes; 0=no)  \
0                                                1.0                         
1                                                1.0                         
2                                                1.0                         
3                                                1.0                         
4                                                1.0                         

   A woman can get a job in the same way as a man (1=yes; 0=no)  \
0                                                1.0              
1                                                1.0              
2                                                1.0              
3                                                1.0              
4                                                1.0              

   A woman can obtain a judgment of divorce in the same way as a man (1=yes; 0=no)  \
0                                                1.0                                 
1                                                1.0                                 
2                                                1.0                                 
3                                                1.0                                 
4                                                1.0                                 

   A woman can open a bank account in the same way as a man (1=yes; 0=no)  \
0                                                1.0                        
1                                                1.0                        
2                                                1.0                        
3                                                1.0                        
4                                                1.0                        

   A woman can register a business in the same way as a man (1=yes; 0=no)  \
0                                                1.0                        
1                                                1.0                        
2                                                1.0                        
3                                                1.0                        
4                                                1.0                        

   A woman can sign a contract in the same way as a man (1=yes; 0=no)  \
0                                                0.0                    
1                                                0.0                    
2                                                1.0                    
3                                                1.0                    
4                                                1.0                    

   A woman can travel outside her home in the same way as a man (1=yes; 0=no)  \
0                                                1.0                            
1                                                1.0                            
2                                                1.0                            
3                                                1.0                            
4                                                1.0                            

   ...  Country Name_Ukraine  Country Name_United Arab Emirates  \
0  ...                     0                                  0   
1  ...              

In [3]:
# # create X and y dataframes
X = df.drop(columns=['Year', 'Share of STEM Graduates'])  # Features except Year and Share of STEM Graduates
y = df['Share of STEM Graduates']  # Target variable (proportion of female STEM graduates)

In [4]:
# #split data into train and test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=17)
y_test

396   -0.724499
227    0.373087
673    0.698001
702    0.690436
643   -0.198359
         ...   
35    -0.187206
870   -2.901920
503   -0.406601
472    0.413398
211    0.321537
Name: Share of STEM Graduates, Length: 215, dtype: float64

## Choosing the Best parameters

The below code was not running on my computer and causing it to shutdown multiple times. So, after talking to Meghan, our TA, we are commenting out the feature seletion code and creating a csv file with the best features ran by my teammate.

In [5]:
# #initialize sfs, use linear regression as the baseline model for this question
# sfs = SequentialFeatureSelector(estimator = LinearRegression(),
#                                 n_features_to_select = "auto",
#                                 direction = 'forward',
#                                 scoring = 'neg_mean_squared_error',
#                                 cv = 2)

# #fit the data to sfs
# sfs = sfs.fit(X_train, y_train)

# # #retrieve the and print the names of the selected features
# # feature_names = np.array(df.columns.difference(['Share of STEM Graduates']))
# # selected_feature_names = feature_names[sfs.get_support()].tolist()
# # print("Selected features:", selected_feature_names)

# # # transform X_train and X_test to include only the selected features
# X_train_selected = sfs.transform(X_train)
# X_test_selected = sfs.transform(X_test)

# display the shape of transformed X_train_selected and X_test_selected
# print("Transformed X_train shape:", X_train_selected.shape)
# print("Transformed X_test shape:", X_test_selected.shape)

In [6]:
#X_train_selected_df = pd.DataFrame(X_train_selected, columns = selected_feature_names)
#X_train_selected_df.to_csv('X_train_selected.csv')

In [7]:
#X_test_selected_df = pd.DataFrame(X_test_selected, columns = selected_feature_names)
#X_test_selected_df.to_csv('X_test_selected.csv')

In [8]:
X_train_selected = pd.read_csv('X_train_selected.csv')
X_test_selected = pd.read_csv('X_test_selected.csv')
y_train = pd.read_csv('y_train.csv', index_col=0)
y_test = pd.read_csv('y_test.csv', index_col=0)

y_train
y_train.shape

(856, 1)

In [9]:
# Check the shapes of the training and testing sets to ensure consistency
print("Training set shape:", X_train.shape, y_train.shape)
print("Testing set shape:", X_test.shape, y_test.shape)
# X_train_selected_array = X_train_selected.to_numpy()
# X_train_selected_array
# X_test_selected_array = X_test_selected.to_numpy()
# X_test_selected_array


Training set shape: (856, 187) (856, 1)
Testing set shape: (215, 187) (215, 1)


## Initial Baseline using Linear Regression

In [10]:

# Defining features and target variable
X = df.drop(columns=['Year', 'Share of STEM Graduates'])  # Features except Year and Share of STEM Graduates
y = df['Share of STEM Graduates']  # Target variable (proportion of female STEM graduates)

# Splitting the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Creating and training the linear regression model
model = LinearRegression()
model.fit(X_train, y_train)

#making predictions on the training and test sets
y_pred_train = model.predict(X_train)
y_pred_test = model.predict(X_test)

#evaluate the model using MSE and R squared
mse_train = mean_squared_error(y_train,y_pred_train)
mse_test = mean_squared_error(y_test,y_pred_test)
r_sq_train = model.score(X_train, y_train)
r_sq_test = model.score(X_test, y_test)

#print the MSE and R_squared
print('MSE (Train): ', round(mse_train, 3))
print('MSE (Test): ', round(mse_test, 3))
print('R-Squared (Train): ', round(r_sq_train, 3))
print('R-Squared (Test): ', round(r_sq_test, 3))

# Getting the coefficients (indicating the importance of each feature)
coefficients = pd.DataFrame({'Feature': X.columns, 'Coefficient': model.coef_})
print("Coefficients:")
print(coefficients)


MSE (Train):  0.134
MSE (Test):  0.222
R-Squared (Train):  0.828
R-Squared (Test):  0.68
Coefficients:
                                               Feature  Coefficient
0                                           Unnamed: 0    -0.007898
1    A woman can be head of household in the same w...    -0.327750
2    A woman can choose where to live in the same w...    -1.672310
3    A woman can get a job in the same way as a man...     0.142569
4    A woman can obtain a judgment of divorce in th...    -0.159842
..                                                 ...          ...
182                            Country Name_Uzbekistan     7.444480
183                         Country Name_Venezuela, RB     6.494867
184                              Country Name_Viet Nam     7.757411
185                    Country Name_West Bank and Gaza     4.765331
186                              Country Name_Zimbabwe     8.858079

[187 rows x 2 columns]


The initial model performs reasonably well with an r-squared of 0.799 for the training data and 0.522 for the test data. However, the lower test r-squared compared to the training r-squared suggests a potential issue with overfitting. Furthermore, the mean squared error (MSE) of the test data is significantly higher than that of the training data. While this basic linear regression model shows satisfactory performance, exploring more sophisticated models will provide insights into potential improvements.

## Defining the function for model selection

In [11]:
def question1(model_type, X_train, y_train, X_test, y_test):

    # Step 1: Initialize a model object
    if model_type == 'Linear Regression':
      model = LinearRegression()
    elif model_type == 'Ridge':
      model = Ridge()
    elif model_type == 'Random Forest':
      model = RandomForestRegressor()
    elif model_type == 'SVM':
      model = svm.LinearSVR()
    else:
      print('Please specify the model type')

    # Step 2: Train the model
    model.fit(X_train, y_train)

    # Step 3: Make predictions
    y_pred_train = model.predict(X_train)
    y_pred_test = model.predict(X_test)

    # Step 4: Evaluate the model performance

    ## MSE
    #evaluate the model using MSE and R squared
    mse_train = mean_squared_error(y_train,y_pred_train)
    mse_test = mean_squared_error(y_test,y_pred_test)
    r_sq_train = model.score(X_train, y_train)
    r_sq_test = model.score(X_test, y_test)

    #print the MSE and R_squared
    print('MSE (Train): ', round(mse_train, 3))
    print('MSE (Test): ', round(mse_test, 3))
    print('R-Squared (Train): ', round(r_sq_train, 3))
    print('R-Squared (Test): ', round(r_sq_test, 3))

    return mse_train, mse_test, r_sq_train, r_sq_test

### Model 1: Linear Regression

In [18]:
#running the model with the selected features
print('Linear Regression:')
mse_train_lr, mse_test_lr, r_sq_train_lr, r_sq_test_lr = question1('Linear Regression', X_train_selected, y_train, X_test_selected, y_test)

Linear Regression:
MSE (Train):  0.68
MSE (Test):  1.453052901042815e+18
R-Squared (Train):  0.128
R-Squared (Test):  -2.0914972599417257e+18


### Model 2: Ridge

In [26]:
# Define the alpha parameter for Ridge regularization
#running the ridge model with the selected features
print('Ridge Model:')
mse_train_ridge, mse_test_ridge, r_sq_train_ridge, r_sq_test_ridge = question1('Ridge', X_train, y_train, X_test, y_test)
print('\n')

Ridge Model:
MSE (Train):  0.173
MSE (Test):  0.267
R-Squared (Train):  0.778
R-Squared (Test):  0.616




### Model 3: Random Forrest

In [32]:
#reshape the target variable
y_train_ensemble = y_train.values.ravel()
y_test_ensemble = y_test.values.ravel()

#running the random forest model with the selected features
print('Random Forest Regressor:')
mse_train_rf, mse_test_rf, r_sq_train_rf, r_sq_test_rf = question1('Random Forest', X_train_selected, y_train_ensemble, X_test_selected, y_test_ensemble)
print('\n')


Random Forest Regressor:
MSE (Train):  0.12
MSE (Test):  0.793
R-Squared (Train):  0.846
R-Squared (Test):  -0.141




### Model 4:  SVM

In [16]:
#running the elasticnet model with the selected features
print('SVM model:')
mse_train_svm, mse_test_svm, r_sq_train_svm, r_sq_test_svm = question1('SVM', X_train, y_train.values.ravel(), X_test, y_test.values.ravel())

SVM model:
MSE (Train):  5.166
MSE (Test):  4.969
R-Squared (Train):  -5.626
R-Squared (Test):  -6.152


/Users/sharonjohn/anaconda3/lib/python3.10/site-packages/sklearn/svm/_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


## Dataframe printing out all the models results

In [17]:
# Create a dictionary to hold evauluation metrics from the models
data = {

    'Model': [
        'Linear Regression',
        'Ridge',
        'Random Forest',
        'SVM'
        ],


    'R Squared Train': [
        r_sq_train_lr,
        r_sq_train_ridge,
        r_sq_train_rf,
        r_sq_train_svm
        ],

    'R Squared Test': [
        r_sq_test_lr,
        r_sq_test_ridge,
        r_sq_test_rf,
        r_sq_test_svm
        ],

    'MSE Train': [
        mse_train_lr,
        mse_train_ridge,
        mse_train_rf,
        mse_train_svm
        ],
    
    
    'MSE Test': [
        mse_test_lr,
        mse_test_ridge,
        mse_test_rf,
        mse_test_svm
    ]
    }

# Create and display a dataframe
results_df = pd.DataFrame(data)
results_df

Model  R Squared Train  R Squared Test  MSE Train      MSE Test
0  Linear Regression         0.127546   -2.091497e+18   0.680263  1.453053e+18
1              Ridge         0.120696   -1.568315e-01   0.685604  8.037005e-01
2      Random Forest         0.845023   -1.781536e-01   0.120838  8.185139e-01
3                SVM        -5.626141   -6.152021e+00   5.166481  4.968816e+00

Result: None of the models performed well with the selected features suprisingly. But from all of them Ridge performed the best despite the overfitting

## Tuning for hyperparameters

### Model 1: Linear Regression 

There is no use for hypertuning linear regressions as it does not provide any useful parameters to tune

### Model 2: Ridge

In [20]:
#Testing for random alphas
# Generate a random list of regularization parameters (alphas)
# Smaller alphas --> weaker regularization, higher alphas --> stronger regularization
alphas = np.logspace(-2, 8, 100)

# Create a list where you will store the information
model_ridge_coefficients = []
mse_train_ridge_list = []
mse_test_ridge_list = []

# Try each set of alphas in your Ridge model
for a in alphas:
    #Pick the model type by initializing a model object and set the alpha from the list as a model parameter
    model_ridge = Ridge()
    model_ridge.set_params(alpha=a)

    #Train the model by passing some data
    model_ridge.fit(X_train_selected, y_train)

    # Get predictions
    y_pred_train_ridge_experiment = model_ridge.predict(X_train_selected)
    y_pred_test_ridge_experiment = model_ridge.predict(X_test_selected)

    #Evaluate the model performance
    mse_train_ridge = mean_squared_error(y_train, y_pred_train_ridge_experiment)
    mse_test_ridge = mean_squared_error(y_test, y_pred_test_ridge_experiment)

    # Append all results to the lists so we could look at them and also plot them
    model_ridge_coefficients.append(model_ridge.coef_)
    mse_train_ridge_list.append(mse_train_ridge)
    mse_test_ridge_list.append(mse_test_ridge)


In [21]:
#Find and print the best alpha

# Find the lowest value of MSE test value
min_mse_test_ridge = np.min(mse_test_ridge_list)
print("Min MSE Test: ", min_mse_test_ridge)

# Find the index of the lowest MSE test value (use np.argmin)
min_mse_test_ridge_index = np.argmin(mse_test_ridge_list)
print("Index of Min MSE test: ", min_mse_test_ridge_index)

# Call alphas with the index of the lowest MSE test value
best_alpha = alphas[min_mse_test_ridge_index]
print("Best alpha: ", best_alpha)

Min MSE Test:  0.6950633547170354
Index of Min MSE test:  99
Best alpha:  100000000.0


### Picking the best alpha in the model

In [22]:
# Pick the model type by initializing a model object and set the alpha to the best alpha
model_ridge_best_alpha = Ridge()
model_ridge_best_alpha.set_params(alpha=best_alpha)

# Train the model by passing some data
model_ridge_best_alpha.fit(X_train_selected, y_train)

# Get predictions
y_pred_train_ridge_best_alpha = model_ridge_best_alpha.predict(X_train_selected)
y_pred_test_ridge_best_alpha = model_ridge_best_alpha.predict(X_test_selected)

# Evaluate the model performance
r_squared_train_ridge_best_alpha = model_ridge_best_alpha.score(X_train_selected, y_train)
r_squared_test_ridge_best_alpha = model_ridge_best_alpha.score(X_test_selected, y_test)
mse_train_ridge_best_alpha = mean_squared_error(y_train, y_pred_train_ridge_best_alpha)
mse_test_ridge_best_alpha = mean_squared_error(y_test, y_pred_test_ridge_best_alpha)

print("Best Ridge Train R^2: ", round(r_squared_train_ridge_best_alpha, 3))
print("Best Ridge Test R^2: ", round(r_squared_test_ridge_best_alpha, 3))
print("Best Ridge Train MSE: ", round(mse_train_ridge_best_alpha, 3))
print("Best Ridge Test MSE: ", round(mse_test_ridge_best_alpha, 3))

Best Ridge Train R^2:  0.002
Best Ridge Test R^2:  -0.0
Best Ridge Train MSE:  0.778
Best Ridge Test MSE:  0.695


### Model 3: Random Forrest

In [23]:
# Put the parameters options
parameters = {'max_depth': [5, 10, 20],
              'max_features': [5, 10, 20],
              'min_samples_split': [2, 5, 10]}

# Initialize the model
rfr = RandomForestRegressor()

#Initialize a GridSearchCV object by passing your model, parameters
model_grid_search = GridSearchCV(rfr, parameters, cv=5)

#Train the GridSearchCV
model_grid_search.fit(X_train_selected, y_train_ensemble)

##Print the best parameters
print("Best Parameters:", model_grid_search.best_params_)

## Save the best model into a separate variable to be used later
rfr_finetuned = model_grid_search.best_estimator_

#Evaluate the model

# Get predictions
y_pred_train_rfr_finetuned = rfr_finetuned.predict(X_train_selected)
y_pred_test_rfr_finetuned = rfr_finetuned.predict(X_test_selected)

## r-squared
r_squared_train_rfr_finetuned = rfr_finetuned.score(X_train_selected, y_train_ensemble)
r_squared_test_rfr_finetuned = rfr_finetuned.score(X_test_selected, y_test_ensemble)

## mse
mse_train_rfr_finetuned = mean_squared_error(y_train_ensemble, y_pred_train_rfr_finetuned)
mse_test_rfr_finetuned = mean_squared_error(y_test_ensemble, y_pred_test_rfr_finetuned)

print('R-squared (Train):', round(r_squared_train_rfr_finetuned, 3))
print('R-squared (Test):', round(r_squared_test_rfr_finetuned, 3))
print("Finetuned Ridge MSE (Train): ", round(mse_train_rfr_finetuned, 3))
print("Finetuned Ridge MSE (Test): ", round(mse_test_rfr_finetuned, 3))

Best Parameters: {'max_depth': 5, 'max_features': 5, 'min_samples_split': 5}
R-squared (Train): 0.079
R-squared (Test): -0.005
Finetuned Ridge MSE (Train):  0.718
Finetuned Ridge MSE (Test):  0.698


### Model 4: SVR

In [25]:
from sklearn.svm import SVR
from sklearn.model_selection import GridSearchCV

# Define the parameters to search over
parameters_svm = {'C': [0.1, 1],  # Regularization parameter
                  'kernel': ['linear', 'rbf', 'poly']}  # Kernel types

# Initialize the SVR model
svr = SVR()

# Initialize a GridSearchCV object by passing the SVR model and parameters
grid_search_svm = GridSearchCV(svr, parameters_svm, cv=5)

# Train the GridSearchCV
grid_search_svm.fit(X_train_selected, y_train.values.ravel())

# Print the best parameters
print("Best Parameters:", grid_search_svm.best_params_)

# Save the best model into a separate variable to be used later
best_svm_model = grid_search_svm.best_estimator_

#evaluating the model

# Get predictions
y_pred_train_svm = best_svm_model.predict(X_train_selected)
y_pred_test_svm = best_svm_model.predict(X_test_selected)

# Calculate R-squared
best_r_squared_train_svm = best_svm_model.score(X_train_selected, y_train)
best_r_squared_test_svm = best_svm_model.score(X_test_selected, y_test)

# Calculate MSE
best_mse_train_svm = mean_squared_error(y_train, y_pred_train_svm)
best_mse_test_svm = mean_squared_error(y_test, y_pred_test_svm)



Best Parameters: {'C': 0.1, 'kernel': 'poly'}


NameError: name 'r_squared_train_svm' is not defined

In [28]:
# Print evaluation metrics
print('R-squared (Train):', round(best_r_squared_train_svm, 3))
print('R-squared (Test):', round(best_r_squared_test_svm, 3))
print("SVR MSE (Train): ", round(best_mse_train_svm, 3))
print("SVR MSE (Test): ", round(best_mse_test_svm, 3))

R-squared (Train): 0.01
R-squared (Test): -0.021
SVR MSE (Train):  0.772
SVR MSE (Test):  0.709


## Dataframe printing out all the models results

In [31]:
# Create a dictionary to hold evauluation metrics from the models
finetuned_data = {

    'Model': [
        'Linear Regression',
        'Ridge',
        'Ridge (Hypertuned)',
        'Random Forest',
        'Random Forest (Hypertuned)',
        'SVR',
        'SVR (Hypertuned)'
        ],


    'R Squared Train': [
        r_sq_train_lr,
        r_sq_train_ridge,
        r_squared_train_ridge_best_alpha,
        r_sq_train_rf,
        r_squared_train_rfr_finetuned,
        r_sq_train_svm,
        best_r_squared_train_svm 
        ],

    'R Squared Test': [
        r_sq_test_lr,
        r_sq_test_ridge,
        r_squared_test_ridge_best_alpha,
        r_sq_test_rf,
        r_squared_test_rfr_finetuned,
        r_sq_train_svm,
        best_r_squared_test_svm 
        ],

    'MSE Train': [
        mse_train_lr,
        mse_train_ridge,
        mse_train_ridge_best_alpha,
        mse_train_rf,
        mse_train_rfr_finetuned,
        mse_train_svm,
        best_mse_train_svm
        ],

    'MSE Test': [
        mse_test_lr,
        mse_test_ridge,
        mse_test_ridge_best_alpha,
        mse_test_rf,
        mse_test_rfr_finetuned,
        mse_test_svm,
        mse_test_svm
    ]
    }

# Create and display a dataframe
finetuned_results_df = pd.DataFrame(finetuned_data)
finetuned_results_df

Model  R Squared Train  R Squared Test  MSE Train  \
0           Linear Regression         0.127546   -2.091497e+18   0.680263   
1                       Ridge         0.778296    6.157175e-01   0.172865   
2          Ridge (Hypertuned)         0.001846   -4.612364e-04   0.778272   
3               Random Forest         0.845023   -1.781536e-01   0.120838   
4  Random Forest (Hypertuned)         0.078816   -4.971439e-03   0.718258   
5                         SVR        -5.626141   -5.626141e+00   5.166481   
6            SVR (Hypertuned)         0.009599   -2.118498e-02   0.772227   

       MSE Test  
0  1.453053e+18  
1  2.669775e-01  
2  6.950634e-01  
3  8.185139e-01  
4  6.981968e-01  
5  4.968816e+00  
6  4.968816e+00

Observations:
- The Ridge model performed the best overall with a 0.77 R squared train and 6.15 test score.
- The model that improved the most with hyperparameter tuning was the SVR model.
- Overall, the other models did not perform well with concerns of overfitting. 
- Another observaiton to note is that suprisingly the features selection parameters during feautre engineering did not make the models perform better. This may be due to the fact that the data was not handled properly and I will have to check on what happened there. 

### References

1. Lab 3 Solutions
2. Lab 6 Solutions
3. https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.GradientBoostingRegressor.html
https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html
https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html
4. Lab 8 Solutions
5. https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVR.html
6. HW1
